### ISSUE TITLE
Apple M1 issue with 'rebar3 eunit' and 'rebar3 shell' commands

### ISSUE THREAD

**Issue by brucify** [rusterlium/rebar3_cargo#9]
Maybe an Apple M1 issue, works fine on Intel-based mac and ubuntu. 

    aarch64-apple-darwin21.6.0
    rustc 1.65.0 (897e37553 2022-11-02)
    rebar 3.20.0 on Erlang/OTP 25 Erts 13.1.1
    macOS 12.6

The `rebar3 eunit` and `rebar3 shell` commands would crash each other, if run after one another.

To reproduce using a test repo https://github.com/brucify/my_nif:

### Scenario 1 (first `rebar3 eunit` then `rebar3 shell`):
First do `rebar3 eunit` and it works.
```
$ git clone https://github.com/brucify/my_nif.git
$ cd my_nif
$ rebar3 eunit
===> Fetching rebar3_cargo v0.1.0
===> Fetching cargo v0.1.2
===> Fetching jsx v2.11.0
===> Analyzing applications...
===> Compiling jsx
===> Compiling cargo
===> Compiling rebar3_cargo
===> Verifying dependencies...
    Updating crates.io index
   Compiling memchr v2.5.0
   Compiling proc-macro2 v1.0.47
   Compiling regex-syntax v0.6.28
   Compiling quote v1.0.21
   Compiling unicode-ident v1.0.5
   Compiling syn v1.0.105
   Compiling void v1.0.2
   Compiling heck v0.4.0
   Compiling rustler v0.26.0
   Compiling lazy_static v1.4.0
   Compiling unreachable v1.0.0
   Compiling aho-corasick v0.7.20
   Compiling regex v1.7.0
   Compiling rustler_sys v2.2.0
   Compiling rustler_codegen v0.26.0
   Compiling my_nif v0.1.0 (/Users/bruce/git/my_nif/native/my_nif)
    Finished dev [unoptimized + debuginfo] target(s) in 7.57s
===>   Copying libmy_nif...
===> Analyzing applications...
===> Compiling my_nif
===> Performing EUnit tests...
.......
Finished in 0.398 seconds
7 tests, 0 failures
   Compiling my_nif v0.1.0 (/Users/bruce/git/my_nif/native/my_nif)
    Finished test [unoptimized + debuginfo] target(s) in 0.31s
     Running unittests src/lib.rs (/Users/bruce/git/my_nif/_build/test/lib/my_nif/target/debug/deps/my_nif-54c8acf2ab3a5deb)
```
But `rebar3 shell` would crash:
```
$ rebar3 shell
===> Verifying dependencies...
   Compiling memchr v2.5.0
   Compiling proc-macro2 v1.0.47
   Compiling regex-syntax v0.6.28
   Compiling quote v1.0.21
   Compiling unicode-ident v1.0.5
   Compiling syn v1.0.105
   Compiling void v1.0.2
   Compiling heck v0.4.0
   Compiling rustler v0.26.0
   Compiling lazy_static v1.4.0
   Compiling unreachable v1.0.0
   Compiling aho-corasick v0.7.20
   Compiling regex v1.7.0
   Compiling rustler_sys v2.2.0
   Compiling rustler_codegen v0.26.0
   Compiling my_nif v0.1.0 (/Users/bruce/git/my_nif/native/my_nif)
    Finished dev [unoptimized + debuginfo] target(s) in 5.88s
===>   Copying libmy_nif...
===> Analyzing applications...
===> Compiling my_nif
Erlang/OTP 25 [erts-13.1.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]

Eshell V13.1.1  (abort with ^G)
1> my_nif:add(1,1).
[1]    85287 killed     rebar3 shell
                                    %
```

### Scenario 2  (first `rebar3 shell` then `rebar3 eunit`):

We reset the repo. First do `rebar3 shell` and it works.
```
$ rm -rf priv/crates && rm -rf _build
$ rebar3 shell
===> Fetching rebar3_cargo v0.1.0
===> Fetching cargo v0.1.2
===> Fetching jsx v2.11.0
===> Analyzing applications...
===> Compiling jsx
===> Compiling cargo
===> Compiling rebar3_cargo
===> Verifying dependencies...
   Compiling memchr v2.5.0
   Compiling proc-macro2 v1.0.47
   Compiling regex-syntax v0.6.28
   Compiling quote v1.0.21
   Compiling unicode-ident v1.0.5
   Compiling syn v1.0.105
   Compiling void v1.0.2
   Compiling heck v0.4.0
   Compiling rustler v0.26.0
   Compiling lazy_static v1.4.0
   Compiling unreachable v1.0.0
   Compiling aho-corasick v0.7.20
   Compiling regex v1.7.0
   Compiling rustler_sys v2.2.0
   Compiling rustler_codegen v0.26.0
   Compiling my_nif v0.1.0 (/Users/bruce/git/my_nif/native/my_nif)
    Finished dev [unoptimized + debuginfo] target(s) in 5.75s
===>   Copying libmy_nif...
===> Analyzing applications...
===> Compiling my_nif
Erlang/OTP 25 [erts-13.1.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]

Eshell V13.1.1  (abort with ^G)
1> my_nif:add(1,1).
2
2>
```
`Ctrl+C` out. Now `rebar3 eunit` would crash:
```
$ rebar3 eunit
===> Verifying dependencies...
   Compiling memchr v2.5.0
   Compiling proc-macro2 v1.0.47
   Compiling quote v1.0.21
   Compiling unicode-ident v1.0.5
   Compiling regex-syntax v0.6.28
   Compiling syn v1.0.105
   Compiling void v1.0.2
   Compiling rustler v0.26.0
   Compiling heck v0.4.0
   Compiling lazy_static v1.4.0
   Compiling unreachable v1.0.0
   Compiling aho-corasick v0.7.20
   Compiling regex v1.7.0
   Compiling rustler_sys v2.2.0
   Compiling rustler_codegen v0.26.0
   Compiling my_nif v0.1.0 (/Users/bruce/git/my_nif/native/my_nif)
    Finished dev [unoptimized + debuginfo] target(s) in 5.83s
===>   Copying libmy_nif...
===> Analyzing applications...
===> Compiling my_nif
===> Performing EUnit tests...
[1]    85693 killed     rebar3 eunit
```
Tested in some other `rebar3_cargo` based repos, getting the same crashes.

**Comment by brucify** [comment]
I found a workaround by adding `{eunit, {cargo, clean}}` into `{post, ...` and both `clean` and `build` into `{pre, ...` in `provider_hooks` in rebar.config. 

However this makes `rebar3 eunit` command always compile twice.

```
{provider_hooks, [
    {pre, [ {compile, {cargo, build}}
          , {eunit, {cargo, clean}}
          , {eunit, {cargo, build}}
          ]},
    {post, [ {clean, {cargo, clean}}
           , {eunit, {cargo, test}}
           , {eunit, {cargo, clean}}
           ]}
]}.
```
Another workaround relies on a `Makefile`, by running an extra command `rebar3 cargo clean` every time before both `rebar3 eunit` and `rebar3 shell`

**Comment by belltoy** [comment]
The latest `cargo v0.1.3` should fix this.

**Comment by brucify** [comment]
> The latest `cargo v0.1.3` should fix this.

 Still the same crashes with 0.1.1

**Comment by belltoy** [comment]
It's weird. I have tried to reproduce the two scenarios you provided on my M1 Mac. They both worked fine.

- macOS 13.0.1
- rustc 1.64.0 (a55dd71d5 2022-09-19)
- rebar 3.20.0 on Erlang/OTP 25 Erts 13.0.3

Hex dependencies:

- `{rebar3_cargo, "0.1.1"}`
- `{cargo, "0.1.3"}`

**Comment by brucify** [comment]
Seems to be something wrong with the compiled `.so` file at `priv/crates/my_nif/my_nif.so`

If I remove it before each `rebar3 eunit` and `rebar3 shell` it works fine.

```erlang
file:delete(OutPath),
```

https://github.com/rusterlium/rebar3_cargo/blob/bcddc293ab0584da4ef4b05e0414370acc8fcaba/src/rebar3_cargo_compile.erl#L123-L125

**Comment by evnu** [comment]
We encountered the same (EDIT: or similar?) problem some time ago in `rustler`, see [the code that handles the artifacts](https://github.com/rusterlium/rustler/blob/3a7aa47af95ff7ec616222214cce8d4e443882e9/rustler_mix/lib/rustler/compiler.ex#L192).

Some background information: `cp` is a non-atomic write. Thus, if an Erlang node already has that shared lib loaded, its contents are overwritten and in an inconsistent state until the copy is finished. By deleting the shared libary, the loaded lib is still available to the running node (the file is not immediately deleted, but it cannot be found in the filesystem any longer), and `cp` can create it anew.

**Comment by brucify** [comment]
Fixed by #10. Closing this.
